### PR TITLE
Introduce explicit build mode and clean up code

### LINKED ISSUE #62: GHA: build master w/ deps' masters, not releases (again)
## Blocked by

* #61

### LINKED ISSUE #78: icinga/icingaweb2:master css build failed
In the current master on docker hub the build for icinga.min.css seems to have failed and instead returns this error:

```less
.appearance is undefined in anonymous-file-0.less
#0 /usr/share/icingaweb2/library/vendor/lessphp/lib/Less/Tree/Ruleset.php(143): Less_Tree_Mixin_Call->compile()
#1 /usr/share/icingaweb2/library/vendor/lessphp/lib/Less/Tree/Ruleset.php(86): Less_Tree_Ruleset->EvalMixinCalls()
#2 /usr/share/icingaweb2/library/vendor/lessphp/lib/Less/Tree/Ruleset.php(91): Less_Tree_Ruleset->compile()
#3 /usr/share/icingaweb2/library/vendor/lessphp/lib/Less/Tree/Ruleset.php(91): Less_Tree_Ruleset->compile()
#4 /usr/share/icingaweb2/library/vendor/lessphp/lib/Less/Parser.php(185): Less_Tree_Ruleset->compile()
#5 /usr/share/icingaweb2/library/vendor/lessphp/lessc.inc.php(130): Less_Parser->getCss()
#6 /usr/share/icingaweb2/library/Icinga/Web/LessCompiler.php(230): lessc->compile()
#7 /usr/share/icingaweb2/library/Icinga/Web/StyleSheet.php(207): Icinga\Web\LessCompiler->render()
#8 /usr/share/icingaweb2/library/Icinga/Web/StyleSheet.php(247): Icinga\Web\StyleSheet->render()
#9 /usr/share/icingaweb2/library/Icinga/Application/webrouter.php(64): Icinga\Web\StyleSheet::send()
#10 /usr/share/icingaweb2/public/index.php(4): require_once('/usr/share/icin...')
#11 {main}


.. em;
       width: 1.5em;
     }
   }
   
   .service-grid-link {
     .bg-stateful();
     .rounded-corners();
   
     display: inline-block;
     height: 1.5em;
     vertical-align: middle;
     width: 1.5em;
   }
   }/* Icinga Web 2 | (c) 2021 Icinga GmbH | GPLv2+ */
   
   @enable-color-preference: 999999px;
   @prefer-light-color-scheme: 999999px;
   
..
```

Do you need other input to track this down?

### PR DESCRIPTION
**Pull request by julianbrost**
This commit introduces an explicit build mode in get-mods.sh that determines how the versions of external modules are selected.

Possible modes are:
 - `release`: chooses latest release for each module
 - `snapshot`: uses a snapshot/development version for each module (if available)

In `action.bash`, the mode is automatically set based on the `GITHUB_REF` variable. For tags starting with `v` (i.e. version tags like `v2.9.0`) it is `release`, other tag names are not supported. For branches it is `snapshot`.

#### master
https://github.com/julianbrost/icingaweb2/runs/5550768750?check_suite_focus=true

Uses branch name as tag for the Docker image, selects snapshot mode and selects dev versions:

```
 + MODE=snapshot
+ TAG=master
+ TAG=master
+ mkimg
+ test -n master
+ test -n snapshot
[...]
+ REF=snapshot/nightly
```

#### v42.23
https://github.com/julianbrost/icingaweb2/runs/5550770184?check_suite_focus=true

Uses tag name with `v` prefix trimmed as tag for the Docker image, selects release mode and selects released versions:

```
+ MODE=release
+ TAG=42.23
+ mkimg
+ test -n 42.23
+ test -n release
[...]
+ REF=v0.7.0
```

#### action-test-without-prefix
https://github.com/julianbrost/icingaweb2/runs/5550850774?check_suite_focus=true

Uses branch name as tag for the Docker image, selects snapshot mode and selects dev versions:

```
 + MODE=snapshot
+ TAG=action-test-without-prefix
+ TAG=action-test-without-prefix
+ mkimg
+ test -n action-test-without-prefix
+ test -n snapshot
[...]
+ REF=snapshot/nightly
```

#### feature/action-test-with-prefix
https://github.com/julianbrost/icingaweb2/runs/5550852672?check_suite_focus=true

Uses branch name with `feature/` prefix trimmed as tag for the Docker image, selects snapshot mode and selects dev versions:

```
+ MODE=snapshot
+ TAG=feature/action-test-with-prefix
+ TAG=action-test-with-prefix
+ mkimg
+ test -n action-test-with-prefix
+ test -n snapshot
[...]
+ REF=snapshot/nightly
```

fixes #62
closes #73
closes #78

### CODE REVIEW

File: action.bash [hunk 1]
```diff
@@ -31,39 +32,24 @@ push () {
...
-		[[ "$GITHUB_REF" =~ $re_docker_tag ]]
-		if [ -n "${BASH_REMATCH[3]}" ]; then
-			TAG="${BASH_REMATCH[3]}"
-		else
-			TAG="${BASH_REMATCH[2]}"
-		fi
+		case "$GITHUB_REF" in
+			refs/tags/v*)
+				MODE=release
+				TAG=${GITHUB_REF#refs/tags/v}
```
Comment (Al2Klimov) [inline action.bash:40]: At least my PR...

File: action.bash [hunk 2]
```diff
@@ -31,39 +32,24 @@ push () {
...
+				;;
+			refs/heads/*)
+				MODE=snapshot
+				TAG=${GITHUB_REF#refs/heads/}
```
Comment (Al2Klimov) [inline action.bash:44]: ... **D**oesn’t **R**epeat **Y**ourself.

File: get-mods.sh [hunk 3]
```diff
@@ -2,7 +2,21 @@
 # Icinga Web 2 Docker image | (c) 2020 Icinga GmbH | GPLv2+
 set -exo pipefail
 
-BRANCH="$1"
+usage() {
+	cat >&2 <<EOF
+Usage: get-mods.sh <release|snapshot>
+
+  release:  download the latest release version for all external modules
+  snapshot: download a snapshot/development version for all external modules
```
Comment (Al2Klimov) [inline get-mods.sh]: "for"? I'd say "of".

File: get-mods.sh [hunk 4]
```diff
@@ -2,7 +2,21 @@
...
+EOF
+	exit 1
```
Comment (Al2Klimov) [inline get-mods.sh:12]: 1? Not 2?

File: get-mods.sh [hunk 5]
```diff
@@ -2,7 +2,21 @@
...
+}
+
+MODE="$1"
+case "$MODE" in
+	release|snapshot) ;;
+	*) usage ;;
+esac
```
Comment (Al2Klimov) [inline get-mods.sh]: I'd prefer not to mix functions with free standing code.

**Review by Al2Klimov — APPROVED** [review]
Approving as I'm not going to "be a frog" (this kind of program isn’t worth that).

File: action.bash (hunk 2 above)
Comment (julianbrost) [reply action.bash:44]: I'm not sure what repetition you're referring to exactly: two cases that set `MODE` and `TAG` or that each case repeats the string `refs/...`.

File: get-mods.sh (hunk 3 above)
Comment (julianbrost) [reply get-mods.sh]: "for" as in for loop. But yes, "of" is probably better in natural language here.

File: get-mods.sh (hunk 4 above)
Comment (julianbrost) [reply get-mods.sh:12]: Primarily non-zero, nothing will probably check for anything else.

File: action.bash (hunk 2 above)
Comment (Al2Klimov) [reply action.bash:44]: The latter. But... forget it. It's not too important.

**Comment by dwt** [comment]
@julianbrost @lippserd As far as I can tell the docker image is still not usable because of css build failures that this pull request is probably addressing.

a) am I right?
b) if that its the case, what is currently holding this back? Getting the build actually working again would seem to be of rather big importance to me?

**Review by lippserd — APPROVED** [review]

**Comment by dwt** [comment]
🎉

**Pull request merged**